File: 37. Sudoku Solver/37. Sudoku Solver.ts
function solveSudoku(board: string[][]): void {
    solve(board);
}

function isValid(board: string[][], row: number, col: number, num: string): boolean {
    for (let x = 0; x < 9; x++) {
        if (board[row][x] === num) return false;
        if (board[x][col] === num) return false;
        if (board[3 * Math.floor(row / 3) + Math.floor(x / 3)][3 * Math.floor(col / 3) + x % 3] === num) return false;
    }
    return true;
}

function solve(board: string[][]): boolean {
    for (let row = 0; row < 9; row++) {
        for (let col = 0; col < 9; col++) {
            if (board[row][col] === '.') {
                for (let num = '1'; num <= '9'; num = String.fromCharCode(num.charCodeAt(0) + 1)) {
                    if (isValid(board, row, col, num)) {
                        board[row][col] = num;
                        if (solve(board)) return true;
                        board[row][col] = '.';
                    }
                }
                return false;
            }
        }
    }
    return true;
}

File: 37. Sudoku Solver/37. Sudoku Solver.go
func solveSudoku(board [][]byte)  {
    solve(board)
}

func isValid(board [][]byte, row, col int, num byte) bool {
    for x := 0; x < 9; x++ {
        if board[row][x] == num { return false }
        if board[x][col] == num { return false }
        if board[3*(row/3) + x/3][3*(col/3) + x%3] == num { return false }
    }
    return true
}

func solve(board [][]byte) bool {
    for row := 0; row < 9; row++ {
        for col := 0; col < 9; col++ {
            if board[row][col] == '.' {
                for num := byte('1'); num <= '9'; num++ {
                    if isValid(board, row, col, num) {
                        board[row][col] = num
                        if solve(board) { return true }
                        board[row][col] = '.'
                    }
                }
                return false
            }
        }
    }
    return true
}